app/web_conn.go

```diff
@@ -105,35 +108,54 @@ func (c *WebConn) WritePump() {
 				return
 			}
 
-			var msgBytes []byte
-			if evt, ok := msg.(*model.WebSocketEvent); ok {
-				cpyEvt := &model.WebSocketEvent{}
-				*cpyEvt = *evt
-				cpyEvt.Sequence = c.Sequence
-				msgBytes = []byte(cpyEvt.ToJson())
-				c.Sequence++
-			} else {
-				msgBytes = []byte(msg.ToJson())
+			evt, evtOk := msg.(*model.WebSocketEvent)
+
+			skipSend := false
+			if len(c.Send) >= SEND_SLOW_WARN {
+				// When the pump starts to get slow we'll drop non-critical messages
+				if msg.EventType() == model.WEBSOCKET_EVENT_TYPING || msg.EventType() == model.WEBSOCKET_EVENT_STATUS_CHANGE {
+					l4g.Info(fmt.Sprintf("websocket.slow: dropping message userId=%v type=%v channelId=%v", c.UserId, msg.EventType(), evt.Broadcast.ChannelId))
+					skipSend = true
+				}
 			}
 
-			c.WebSocket.SetWriteDeadline(time.Now().Add(WRITE_WAIT))
-			if err := c.WebSocket.WriteMessage(websocket.TextMessage, msgBytes); err != nil {
-				// browsers will appear as CloseNoStatusReceived
-				if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseNoStatusReceived) {
-					l4g.Debug(fmt.Sprintf("websocket.send: client side closed socket userId=%v", c.UserId))
+			if !skipSend {
+				var msgBytes []byte
+				if evtOk {
+					cpyEvt := &model.WebSocketEvent{}
+					*cpyEvt = *evt
+					cpyEvt.Sequence = c.Sequence
+					msgBytes = []byte(cpyEvt.ToJson())
+					c.Sequence++
 				} else {
-					l4g.Debug(fmt.Sprintf("websocket.send: closing websocket for userId=%v, error=%v", c.UserId, err.Error()))
+					msgBytes = []byte(msg.ToJson())
 				}
 
-				return
-			}
+				if len(c.Send) >= SEND_DEADLOCK_WARN {
+					if evtOk {
+						l4g.Error(fmt.Sprintf("websocket.full: message userId=%v type=%v channelId=%v size=%v", c.UserId, msg.EventType(), evt.Broadcast.ChannelId, len(msg.ToJson())))
+					} else {
+						l4g.Error(fmt.Sprintf("websocket.full: message userId=%v type=%v size=%v", c.UserId, msg.EventType(), len(msg.ToJson())))
+					}
+				}
+
+				c.WebSocket.SetWriteDeadline(time.Now().Add(WRITE_WAIT))
+				if err := c.WebSocket.WriteMessage(websocket.TextMessage, msgBytes); err != nil {
+					// browsers will appear as CloseNoStatusReceived
+					if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseNoStatusReceived) {
+						l4g.Debug(fmt.Sprintf("websocket.send: client side closed socket userId=%v", c.UserId))
+					} else {
+						l4g.Debug(fmt.Sprintf("websocket.send: closing websocket for userId=%v, error=%v", c.UserId, err.Error()))
+					}
+
+					return
+				}
 
-			if msg.EventType() == model.WEBSOCKET_EVENT_POSTED {
 				if einterfaces.GetMetricsInterface() != nil {
-					einterfaces.GetMetricsInterface().IncrementPostBroadcast()
+					go einterfaces.GetMetricsInterface().IncrementWebSocketBroadcast(msg.EventType())
```

Comment:
Actually this is already collected here: https://github.com/mattermost/platform/blob/fb6f2a123c7fefc8d8270e273e9ef4007f88cefd/app/web_hub.go#L96